File: electron/agent/mcp/scanner.ts
/** MCP import scanner: detects servers from other tools' config files (Claude Code, Codex, OpenCode, generic) and normalizes them to Tide's McpServerConfig shape. */
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { createLogger } from '../../logger';
import type { McpServerConfig } from './types';

const log = createLogger('mcp/scanner');

export interface DetectedServer {
  name: string;
  config: McpServerConfig;
  source: string; // display label: "Claude Code", "Codex", etc.
  sourceFile: string; // the file path it came from
}

export interface ScanResult {
  servers: DetectedServer[];
  /** Names already present in Tide's config (so the UI can pre-uncheck them). */
  alreadyImported: string[];
}

/** Scan all known sources for MCP server configs. */
export function scanExternalMcpServers(
  tideConfigPath: string,
): ScanResult {
  const home = os.homedir();
  const detected: DetectedServer[] = [];

  // 1. Claude Code — ~/.claude.json (root-level mcpServers)
  scanJsonFile(
    path.join(home, '.claude.json'),
    'Claude Code',
    (d) => d.mcpServers,
    detected,
  );

  // 1b. Claude Code — ~/.claude/settings.json (mcpServers key)
  scanJsonFile(
    path.join(home, '.claude', 'settings.json'),
    'Claude Code',
    (d) => d.mcpServers,
    detected,
  );

  // 2. Codex CLI — ~/.codex/config.toml ([mcp_servers.*] sections)
  scanCodexToml(
    path.join(home, '.codex', 'config.toml'),
    detected,
  );

  // 3. OpenCode — ~/.config/opencode/opencode.json (mcp key)
  scanJsonFile(
    path.join(home, '.config', 'opencode', 'opencode.json'),
    'OpenCode',
    (d) => d.mcp,
    detected,
  );

  // 4. Generic — ~/.agents/mcp.json
  scanJsonFile(
    path.join(home, '.agents', 'mcp.json'),
    'Generic',
    (d) => {
      // Generic format might use mcpServers wrapper OR flat map
      if (d.mcpServers) return d.mcpServers;
      // If all values look like server configs (have type/command/url), treat as flat
      const vals = Object.values(d);
      if (vals.every((v) => v && typeof v === 'object' && !Array.isArray(v))) return d;
      return {};
    },
    detected,
  );

  // Read Tide's existing config to mark already-imported servers
  let alreadyImported: string[] = [];
  try {
    const tideRaw = fs.readFileSync(tideConfigPath, 'utf-8');
    const tideParsed = JSON.parse(tideRaw);
    alreadyImported = Object.keys(tideParsed);
  } catch {
    // Tide config doesn't exist yet — nothing imported
  }

  // Deduplicate detected servers by name (first source wins per name)
  const seen = new Set<string>();
  const deduped = detected.filter((s) => {
    if (seen.has(s.name)) return false;
    seen.add(s.name);
    return true;
  });

  log.info('scan complete', { detected: deduped.length, alreadyImported: alreadyImported.length });

  return { servers: deduped, alreadyImported };
}

// ─── JSON source scanner ──────────────────────────────────────────────

function scanJsonFile(
  filePath: string,
  sourceLabel: string,
  extract: (data: any) => Record<string, any> | undefined,
  out: DetectedServer[],
): void {
  try {
    const raw = fs.readFileSync(filePath, 'utf-8');
    const parsed = JSON.parse(raw);
    const servers = extract(parsed);
    if (!servers || typeof servers !== 'object') return;

    for (const [name, rawConfig] of Object.entries(servers)) {
      if (!rawConfig || typeof rawConfig !== 'object') continue;
      const normalized = normalizeExternalConfig(rawConfig as Record<string, unknown>);
      if (normalized) {
        out.push({ name, config: normalized, source: sourceLabel, sourceFile: filePath });
      }
    }
  } catch {
    // File doesn't exist or can't be parsed — skip silently
  }
}

// ─── Codex TOML scanner ───────────────────────────────────────────────
// Minimal TOML parser for [mcp_servers.*] sections. We only need to extract
// key=value pairs and inline tables { k = "v" } — not a full TOML parser.

function scanCodexToml(
  filePath: string,
  out: DetectedServer[],
): void {
  try {
    const raw = fs.readFileSync(filePath, 'utf-8');
    const sections = parseTomlMcpServers(raw);
    for (const [name, config] of Object.entries(sections)) {
      const normalized = normalizeExternalConfig(config);
      if (normalized) {
        out.push({ name, config: normalized, source: 'Codex', sourceFile: filePath });
      }
    }
  } catch {
    // File doesn't exist — skip
  }
}

/** Minimal TOML parser for [mcp_servers.NAME] sections (plus env/http_headers sub-tables); not a full TOML implementation. */
function parseTomlMcpServers(toml: string): Record<string, Record<string, unknown>> {
  const result: Record<string, Record<string, unknown>> = {};
  let currentServer: string | null = null;
  let currentSubTable: string | null = null; // 'env' or 'http_headers'

  for (const line of toml.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;

    // [mcp_servers.NAME.env] or [mcp_servers.NAME.http_headers]
    const subMatch = trimmed.match(/^\[mcp_servers\.(\w+)\.(env|http_headers)\]/);
    if (subMatch) {
      currentServer = subMatch[1];
      currentSubTable = subMatch[2];
      if (!result[currentServer]) result[currentServer] = {};
      if (!result[currentServer][currentSubTable]) result[currentServer][currentSubTable] = {};
      continue;
    }

    // [mcp_servers.NAME]
    const serverMatch = trimmed.match(/^\[mcp_servers\.(\w+)\]/);
    if (serverMatch) {
      currentServer = serverMatch[1];
      currentSubTable = null;
      if (!result[currentServer]) result[currentServer] = {};
      continue;
    }

    // Any other [section] — reset context
    if (trimmed.startsWith('[')) {
      currentServer = null;
      currentSubTable = null;
      continue;
    }

    if (!currentServer) continue;

    // key = value
    const eqIdx = trimmed.indexOf('=');
    if (eqIdx < 0) continue;
    const key = trimmed.slice(0, eqIdx).trim();
    const valueRaw = trimmed.slice(eqIdx + 1).trim();

    // Inline table: { K = "V", ... }
    if (valueRaw.startsWith('{')) {
      const inlineTable = parseInlineTomlTable(valueRaw);
      result[currentServer][key] = inlineTable;
      continue;
    }

    // Array: ["a", "b"]
    if (valueRaw.startsWith('[')) {
      result[currentServer][key] = parseTomlArray(valueRaw);
      continue;
    }

    // String: "value"
    const strMatch = valueRaw.match(/^"(.*)"$/);
    if (strMatch) {
      const value = strMatch[1];
      if (currentSubTable === 'env' || currentSubTable === 'http_headers') {
        (result[currentServer][currentSubTable] as Record<string, string>)[key] = value;
      } else {
        result[currentServer][key] = value;
      }
      continue;
    }

    // Bare value (number, bool)
    result[currentServer][key] = valueRaw;
  }

  return result;
}

function parseInlineTomlTable(raw: string): Record<string, string> {
  const result: Record<string, string> = {};
  // Strip outer braces
  const inner = raw.replace(/^\{/, '').replace(/\}$/, '').trim();
  // Split on commas (naive — doesn't handle commas inside values)
  for (const part of inner.split(',')) {
    const eqIdx = part.indexOf('=');
    if (eqIdx < 0) continue;
    const k = part.slice(0, eqIdx).trim();
    const v = part.slice(eqIdx + 1).trim().replace(/^"(.*)"$/, '$1');
    result[k] = v;
  }
  return result;
}

function parseTomlArray(raw: string): string[] {
  const inner = raw.replace(/^\[/, '').replace(/\]$/, '').trim();
  if (!inner) return [];
  return inner.split(',').map((s) => s.trim().replace(/^"(.*)"$/, '$1'));
}

// ─── Normalizer ───────────────────────────────────────────────────────

/** Normalize an external server config to Tide's format, inferring type from `command` (stdio) or `url` (http). */
function normalizeExternalConfig(raw: Record<string, unknown>): McpServerConfig | null {
  const config: Partial<McpServerConfig> = {};

  // Determine type
  if (typeof raw.type === 'string' && ['stdio', 'sse', 'http'].includes(raw.type)) {
    config.type = raw.type as McpServerConfig['type'];
  } else if (typeof raw.command === 'string') {
    config.type = 'stdio';
  } else if (typeof raw.url === 'string') {
    config.type = 'http';
  } else {
    // Can't determine transport — skip
    return null;
  }

  // stdio fields
  if (typeof raw.command === 'string') config.command = raw.command;
  if (Array.isArray(raw.args)) config.args = raw.args.filter((a): a is string => typeof a === 'string');

  // env — might come from `env` key (JSON) or inline TOML table
  if (raw.env && typeof raw.env === 'object' && !Array.isArray(raw.env)) {
    const env: Record<string, string> = {};
    for (const [k, v] of Object.entries(raw.env as Record<string, unknown>)) {
      if (typeof v === 'string') env[k] = v;
    }
    if (Object.keys(env).length > 0) config.env = env;
  }

  // remote fields
  if (typeof raw.url === 'string') config.url = raw.url;

  // Codex uses http_headers instead of headers — we can't map headers to
  // Tide's env-based secret model, so skip them (user can re-add manually)
  // unless the URL has an auth query param already.

  return config as McpServerConfig;
}

File: src-tauri/crates/tide-mcp/src/scanner.rs
//! MCP import scanner — port of `app/core/agent/mcp/scanner.ts` @ 91ec558.
//! Detects servers from other tools' config files (Claude Code, Codex,
//! OpenCode, generic) and normalizes them to Tide's [`McpServerConfig`].
//!
//! Deviation from the TS: the already-imported list comes from the resolved
//! user server map (config.json's `mcpServers`) instead of a raw read of the
//! legacy `mcp.json` — post-migration that file is not where Tide's servers
//! live, and the intent is "names already present in Tide's config".

use std::collections::BTreeMap;
use std::path::Path;

use serde::Serialize;
use serde_json::{Map, Value};

use crate::config::{McpConfigFile, McpServerConfig, McpTransportType};

/// A server detected in another tool's config file — the TS `DetectedServer`.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DetectedServer {
    pub name: String,
    pub config: McpServerConfig,
    /// Display label: "Claude Code", "Codex", etc.
    pub source: String,
    /// The file path it came from.
    pub source_file: String,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ScanResult {
    pub servers: Vec<DetectedServer>,
    /// Names already present in Tide's config (so the UI can pre-uncheck).
    pub already_imported: Vec<String>,
}

/// Scan all known sources under `home` for MCP server configs.
pub fn scan_external_mcp_servers(home: &Path, tide_servers: &McpConfigFile) -> ScanResult {
    let mut detected: Vec<DetectedServer> = Vec::new();

    // 1. Claude Code — ~/.claude.json (root-level mcpServers).
    scan_json_file(
        &home.join(".claude.json"),
        "Claude Code",
        &mut detected,
        extract_mcp_servers,
    );
    // 1b. Claude Code — ~/.claude/settings.json (mcpServers key).
    scan_json_file(
        &home.join(".claude").join("settings.json"),
        "Claude Code",
        &mut detected,
        extract_mcp_servers,
    );
    // 2. Codex CLI — ~/.codex/config.toml ([mcp_servers.*] sections).
    scan_codex_toml(&home.join(".codex").join("config.toml"), &mut detected);
    // 3. OpenCode — ~/.config/opencode/opencode.json (mcp key).
    scan_json_file(
        &home.join(".config").join("opencode").join("opencode.json"),
        "OpenCode",
        &mut detected,
        |d| d.get("mcp").and_then(Value::as_object),
    );
    // 4. Generic — ~/.agents/mcp.json (mcpServers wrapper OR flat map).
    scan_json_file(
        &home.join(".agents").join("mcp.json"),
        "Generic",
        &mut detected,
        extract_generic_servers,
    );

    // Deduplicate by name (first source wins per name).
    let mut seen = std::collections::BTreeSet::new();
    detected.retain(|server| seen.insert(server.name.clone()));

    ScanResult {
        servers: detected,
        already_imported: tide_servers.keys().cloned().collect(),
    }
}

// ─── JSON source scanner ──────────────────────────────────────────────

type Extractor = fn(&Map<String, Value>) -> Option<&Map<String, Value>>;

fn extract_mcp_servers(data: &Map<String, Value>) -> Option<&Map<String, Value>> {
    data.get("mcpServers").and_then(Value::as_object)
}

fn extract_generic_servers(data: &Map<String, Value>) -> Option<&Map<String, Value>> {
    if let Some(servers) = data.get("mcpServers") {
        return servers.as_object();
    }
    // If all values look like server configs (objects), treat as flat.
    if !data.is_empty() && data.values().all(|v| v.is_object()) {
        return Some(data);
    }
    None
}

fn scan_json_file(
    file_path: &Path,
    source_label: &str,
    out: &mut Vec<DetectedServer>,
    extract: Extractor,
) {
    // Missing/unparseable file — skip silently like the TS.
    let Ok(raw) = std::fs::read_to_string(file_path) else {
        return;
    };
    let Ok(parsed) = serde_json::from_str::<Value>(&raw) else {
        return;
    };
    let Some(root) = parsed.as_object() else {
        return;
    };
    let Some(servers) = extract(root) else {
        return;
    };
    for (name, raw_config) in servers {
        let Some(raw_map) = raw_config.as_object() else {
            continue;
        };
        if let Some(config) = normalize_external_config(raw_map) {
            out.push(DetectedServer {
                name: name.clone(),
                config,
                source: source_label.to_owned(),
                source_file: file_path.display().to_string(),
            });
        }
    }
}

// ─── Codex TOML scanner ───────────────────────────────────────────────

/// Minimal TOML reader for `[mcp_servers.NAME]` sections (plus env /
/// http_headers sub-tables) — the TS parser's exact subset, not full TOML.
fn scan_codex_toml(file_path: &Path, out: &mut Vec<DetectedServer>) {
    let Ok(raw) = std::fs::read_to_string(file_path) else {
        return;
    };
    for (name, config) in parse_toml_mcp_servers(&raw) {
        if let Some(normalized) = normalize_external_config(&config) {
            out.push(DetectedServer {
                name,
                config: normalized,
                source: "Codex".to_owned(),
                source_file: file_path.display().to_string(),
            });
        }
    }
}

fn parse_toml_mcp_servers(toml: &str) -> BTreeMap<String, Map<String, Value>> {
    let mut result: BTreeMap<String, Map<String, Value>> = BTreeMap::new();
    let mut current_server: Option<String> = None;
    let mut current_sub_table: Option<String> = None;

    for line in toml.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }

        // [mcp_servers.NAME.env] / [mcp_servers.NAME.http_headers], or the
        // bare [mcp_servers.NAME] — the TS regexes' exact subset.
        if let Some(rest) = trimmed.strip_prefix("[mcp_servers.") {
            if let Some(section) = rest.strip_suffix(']') {
                if let Some((name, sub)) = section.split_once('.') {
                    if sub == "env" || sub == "http_headers" {
                        current_server = Some(name.to_owned());
                        current_sub_table = Some(sub.to_owned());
                        let server = result.entry(name.to_owned()).or_default();
                        server
                            .entry(sub.to_owned())
                            .or_insert_with(|| Value::Object(Map::new()));
                        continue;
                    }
                } else {
                    current_server = Some(section.to_owned());
                    current_sub_table = None;
                    result.entry(section.to_owned()).or_default();
                    continue;
                }
            }
        }

        // Any other [section] — reset context.
        if trimmed.starts_with('[') {
            current_server = None;
            current_sub_table = None;
            continue;
        }

        let Some(server_name) = current_server.clone() else {
            continue;
        };
        let Some((key, value_raw)) = trimmed.split_once('=') else {
            continue;
        };
        let key = key.trim();
        let value_raw = value_raw.trim();
        let server = result.entry(server_name).or_default();

        // Inline table: { K = "V", ... }
        if value_raw.starts_with('{') {
            server.insert(key.to_owned(), Value::Object(parse_inline_toml_table(value_raw)));
            continue;
        }
        // Array: ["a", "b"]
        if value_raw.starts_with('[') {
            let items = parse_toml_array(value_raw);
            server.insert(
                key.to_owned(),
                Value::Array(items.into_iter().map(Value::String).collect()),
            );
            continue;
        }
        // String: "value" — sub-table keys nest under env/http_headers.
        if let Some(value) = value_raw
            .strip_prefix('"')
            .and_then(|v| v.strip_suffix('"'))
        {
            match &current_sub_table {
                Some(sub) if sub == "env" || sub == "http_headers" => {
                    if let Some(sub_map) = server
                        .get(sub)
                        .and_then(Value::as_object)
                        .cloned()
                    {
                        let mut merged = sub_map;
                        merged.insert(key.to_owned(), Value::String(value.to_owned()));
                        server.insert(sub.clone(), Value::Object(merged));
                    }
                }
                _ => {
                    server.insert(key.to_owned(), Value::String(value.to_owned()));
                }
            }
            continue;
        }
        // Bare value (number, bool) — kept as the raw string, TS parity.
        server.insert(key.to_owned(), Value::String(value_raw.to_owned()));
    }

    result
}

fn parse_inline_toml_table(raw: &str) -> Map<String, Value> {
    let mut result = Map::new();
    let inner = raw
        .strip_prefix('{')
        .and_then(|r| r.strip_suffix('}'))
        .unwrap_or(raw)
        .trim();
    // Naive comma split — doesn't handle commas inside values (TS parity).
    for part in inner.split(',') {
        let Some((k, v)) = part.split_once('=') else {
            continue;
        };
        let value = v
            .trim()
            .strip_prefix('"')
            .and_then(|v| v.strip_suffix('"'))
            .unwrap_or(v.trim());
        result.insert(k.trim().to_owned(), Value::String(value.to_owned()));
    }
    result
}

fn parse_toml_array(raw: &str) -> Vec<String> {
    let inner = raw
        .strip_prefix('[')
        .and_then(|r| r.strip_suffix(']'))
        .unwrap_or(raw)
        .trim();
    if inner.is_empty() {
        return Vec::new();
    }
    inner
        .split(',')
        .map(|s| {
            s.trim()
                .strip_prefix('"')
                .and_then(|s| s.strip_suffix('"'))
                .unwrap_or(s.trim())
                .to_owned()
        })
        .collect()
}

// ─── Normalizer ───────────────────────────────────────────────────────

/// Normalize an external server config to Tide's format, inferring type
/// from `command` (stdio) or `url` (http); None when no transport can be
/// determined (the entry is skipped).
fn normalize_external_config(raw: &Map<String, Value>) -> Option<McpServerConfig> {
    let mut config = McpServerConfig::default();

    if let Some(kind) = raw.get("type").and_then(Value::as_str) {
        config.r#type = match kind {
            "stdio" => Some(McpTransportType::Stdio),
            "sse" => Some(McpTransportType::Sse),
            "http" => Some(McpTransportType::Http),
            _ => None,
        };
    }
    if config.r#type.is_none() {
        if raw.get("command").and_then(Value::as_str).is_some() {
            config.r#type = Some(McpTransportType::Stdio);
        } else if raw.get("url").and_then(Value::as_str).is_some() {
            config.r#type = Some(McpTransportType::Http);
        } else {
            return None;
        }
    }

    if let Some(command) = raw.get("command").and_then(Value::as_str) {
        config.command = Some(command.to_owned());
    }
    if let Some(args) = raw.get("args").and_then(Value::as_array) {
        let filtered: Vec<String> = args
            .iter()
            .filter_map(Value::as_str)
            .map(str::to_owned)
            .collect();
        if !filtered.is_empty() {
            config.args = Some(filtered);
        }
    }
    if let Some(env) = raw.get("env").and_then(Value::as_object) {
        let mut resolved = BTreeMap::new();
        for (key, value) in env {
            if let Some(value) = value.as_str() {
                resolved.insert(key.clone(), value.to_owned());
            }
        }
        if !resolved.is_empty() {
            config.env = Some(resolved);
        }
    }
    if let Some(url) = raw.get("url").and_then(Value::as_str) {
        config.url = Some(url.to_owned());
    }
    // Codex `http_headers` has no Tide mapping (TS parity — skipped; the
    // user can re-add headers manually).

    Some(config)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn scan_home() -> tempfile::TempDir {
        tempfile::tempdir().unwrap()
    }

    #[test]
    fn json_sources_are_normalized_and_labeled() {
        let home = scan_home();
        std::fs::write(
            home.path().join(".claude.json"),
            r#"{"mcpServers": {"context7": {"type": "http", "url": "https://c7.mcp/dev"}}}"#,
        )
        .unwrap();
        std::fs::create_dir_all(home.path().join(".config").join("opencode")).unwrap();
        std::fs::write(
            home.path().join(".config").join("opencode").join("opencode.json"),
            r#"{"mcp": {"gh": {"command": "npx", "args": ["-y", "gh-mcp"], "env": {"T": "1"}}}}"#,
        )
        .unwrap();
        let result = scan_external_mcp_servers(home.path(), &McpConfigFile::new());
        assert_eq!(result.servers.len(), 2);
        let c7 = result.servers.iter().find(|s| s.name == "context7").unwrap();
        assert_eq!(c7.source, "Claude Code");
        assert_eq!(c7.config.r#type, Some(McpTransportType::Http));
        assert_eq!(c7.config.url.as_deref(), Some("https://c7.mcp/dev"));
        assert!(c7.source_file.ends_with(".claude.json"));
        let gh = result.servers.iter().find(|s| s.name == "gh").unwrap();
        assert_eq!(gh.source, "OpenCode");
        assert_eq!(gh.config.r#type, Some(McpTransportType::Stdio));
        assert_eq!(gh.config.command.as_deref(), Some("npx"));
        assert_eq!(
            gh.config.args.as_deref(),
            Some(&["-y".to_owned(), "gh-mcp".to_owned()][..])
        );
        assert_eq!(gh.config.env.as_ref().unwrap()["T"], "1");
    }

    #[test]
    fn codex_toml_sections_parse_with_env_subtables() {
        let home = scan_home();
        std::fs::create_dir_all(home.path().join(".codex")).unwrap();
        std::fs::write(
            home.path().join(".codex").join("config.toml"),
            r#"
# comment
[mcp_servers.fetch]
command = "uvx"
args = ["mcp-server-fetch"]

[mcp_servers.fetch.env]
FETCH_TIMEOUT = "30"

[other_section]
key = "ignored"
"#,
        )
        .unwrap();
        let result = scan_external_mcp_servers(home.path(), &McpConfigFile::new());
        assert_eq!(result.servers.len(), 1);
        let fetch = &result.servers[0];
        assert_eq!(fetch.name, "fetch");
        assert_eq!(fetch.source, "Codex");
        assert_eq!(fetch.config.command.as_deref(), Some("uvx"));
        assert_eq!(
            fetch.config.args.as_deref(),
            Some(&["mcp-server-fetch".to_owned()][..])
        );
        assert_eq!(fetch.config.env.as_ref().unwrap()["FETCH_TIMEOUT"], "30");
    }

    #[test]
    fn generic_source_accepts_wrapper_and_flat_maps() {
        let home = scan_home();
        std::fs::create_dir_all(home.path().join(".agents")).unwrap();
        std::fs::write(
            home.path().join(".agents").join("mcp.json"),
            r#"{"a": {"command": "run-a"}, "mcpServers": {"b": {"command": "run-b"}}}"#,
        )
        .unwrap();
        let result = scan_external_mcp_servers(home.path(), &McpConfigFile::new());
        // The wrapper wins per the TS extract order.
        assert_eq!(result.servers.len(), 1);
        assert_eq!(result.servers[0].name, "b");
        assert_eq!(result.servers[0].source, "Generic");
    }

    #[test]
    fn dedup_keeps_first_source_and_marks_already_imported() {
        let home = scan_home();
        std::fs::write(
            home.path().join(".claude.json"),
            r#"{"mcpServers": {"shared": {"command": "from-claude"}}}"#,
        )
        .unwrap();
        std::fs::create_dir_all(home.path().join(".claude")).unwrap();
        std::fs::write(
            home.path().join(".claude").join("settings.json"),
            r#"{"mcpServers": {"shared": {"command": "from-settings"}, "only-settings": {"url": "https://x"}}}"#,
        )
        .unwrap();
        let mut tide = McpConfigFile::new();
        tide.insert(
            "shared".to_owned(),
            serde_json::from_str(r#"{"command": "mine"}"#).unwrap(),
        );
        let result = scan_external_mcp_servers(home.path(), &tide);
        assert_eq!(result.servers.len(), 2);
        let shared = result.servers.iter().find(|s| s.name == "shared").unwrap();
        assert_eq!(shared.config.command.as_deref(), Some("from-claude"));
        assert_eq!(result.already_imported, vec!["shared".to_owned()]);
    }

    #[test]
    fn entries_without_a_transport_are_skipped() {
        let home = scan_home();
        std::fs::write(
            home.path().join(".claude.json"),
            r#"{"mcpServers": {"mystery": {"note": "no command or url"}, "fine": {"command": "ok"}}}"#,
        )
        .unwrap();
        let result = scan_external_mcp_servers(home.path(), &McpConfigFile::new());
        assert_eq!(result.servers.len(), 1);
        assert_eq!(result.servers[0].name, "fine");
    }

    #[test]
    fn missing_sources_scan_clean() {
        let home = scan_home();
        let result = scan_external_mcp_servers(home.path(), &McpConfigFile::new());
        assert!(result.servers.is_empty());
        assert!(result.already_imported.is_empty());
    }
}
